File: tests/Unit/UserTest.php
<?php

namespace Tests\Unit;

use App\DTOs\UserDTO;
use App\Models\User;
use App\Repositories\IUserRepository;
use App\Services\UserService;
use Core\Exceptions\ValidationException;
use PHPUnit\Framework\Attributes\Test;
use PHPUnit\Framework\TestCase;

class UserTest extends TestCase
{

  #[Test]
  public function mustCreateUser(){
    // Arrange
    $dto = new UserDTO(
      id: 1,
      name: "Davi",
      email: "davi@gmail.com",
      password: "davi1234"
    );

    $mockUser = new User(
      id: 1,
      name: $dto->name,
      email: $dto->email,
      password: $dto->password
    );

    $userRepository = $this->createStub(IUserRepository::class);

    $userRepository
      ->method('create')
      ->with($dto)
      ->willReturn($mockUser);

    $userService = new UserService($userRepository);

    // Act
    $result = $userService->create($dto);

    // Assert
    $this->assertEquals($mockUser, $result);
  }

  #[Test]
  public function mustThrowExceptToInvalidName(){
    $this->expectException(ValidationException::class);

    // Arrange
    $dto = new UserDTO(
      id: 1,
      name: "",
      email: "davi@gmail.com",
      password: "davi1234"
    );

    $mockUser = new User(
      id: 1,
      name: $dto->name,
      email: $dto->email,
      password: $dto->password
    );

    $userRepository = $this->createStub(IUserRepository::class);

    $userRepository
      ->method('create')
      ->with($dto)
      ->willReturn($mockUser);

    $userService = new UserService($userRepository);

    // Act
    // Must throw exception 
    $result = $userService->create($dto);
  }
}

File: src/test/java/com/project/todo_api/unit/UserTest.java
package com.project.todo_api.unit;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.when;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import com.project.todo_api.dtos.DtoCreateUser;
import com.project.todo_api.models.User;
import com.project.todo_api.repositories.UserRepository;
import com.project.todo_api.services.UserService;

public class UserTest {

  @Mock
  private UserRepository userRepository;

  @InjectMocks
  private UserService userService;

  @BeforeEach
  public void setUp() {
    MockitoAnnotations.openMocks(this);
  }

  @Test
  public void shouldCreateUser(){
    //Arrange
    String name = "Davi";
    String email = "davi@example.com";
    String password = "Password123!";
    DtoCreateUser dtoCreateUser = new DtoCreateUser(name, email, password);
    User user = new User(dtoCreateUser);

    when(this.userRepository.save(user)).thenReturn(user);

    //Act
    User userReturn = this.userService.createUser(dtoCreateUser);

    //Assert
    assertEquals(userReturn, user);
  }
}
